### PR TITLE
Implement 'open' action

### PR DESCRIPTION
**Pull request by kdvolder**
Relates to https://github.com/Microsoft/vscode-spring-boot-dashboard/issues/6

- On launching app:
  - Enable jmx on dynamivally assigned 'free' port
  - Store jmx port in BootApp attribute
- On 'open'
  - Use JMX port to connect to JMX and determine app's http port.
  - Use 'opn' npm package to open apps main url.

### CODE REVIEW

File: src/Controller.ts [hunk 1]
```diff
@@ -7,6 +7,8 @@ import { BootApp, STATE_RUNNING, STATE_INACTIVE } from "./BootApp";
 import { findJvm } from "@pivotal-tools/jvm-launch-utils";
 import * as path from "path";
 import { readAll } from "./stream-util";
+import opn = require('opn')
```
Comment (andxu) [inline src/Controller.ts]: don't use require, use import instead.

File: src/Controller.ts [hunk 2]
```diff
@@ -38,9 +40,24 @@ export class Controller {
...
+            app.jmxPort = jmxport;
+            let vmArgs = '-Dcom.sun.management.jmxremote ' +
+                `-Dcom.sun.management.jmxremote.port=${jmxport} `+ 
+                '-Dcom.sun.management.jmxremote.authenticate=false ' + 
+                '-Dcom.sun.management.jmxremote.ssl=false ' +
+                '-Djava.rmi.server.hostname=localhost '+ 
+                '-Dspring.application.admin.enabled=true';
+            if (targetConfig.vmArgs) {
+                //TODO: smarter merge? What if user is trying to enable jmx themselves on a specific port they choose, for example?
+                vmArgs = vmArgs + ' ' +targetConfig.vmArgs;
```
Comment (andxu) [inline src/Controller.ts]: +<insert space>targetConfig.vmArgs

File: src/Controller.ts [hunk 3]
```diff
@@ -38,9 +40,24 @@ export class Controller {
...
+            }
             const ok: boolean = await vscode.debug.startDebugging(
                 vscode.workspace.getWorkspaceFolder(vscode.Uri.parse(app.path)),
-                Object.assign({}, targetConfig, { noDebug: !debug })
+                Object.assign({}, targetConfig, { 
+                    noDebug: !debug,
+                    vmArgs: vmArgs
```
Comment (andxu) [inline src/Controller.ts]: vmArgs

File: src/Controller.ts [hunk 4]
```diff
@@ -78,35 +95,28 @@ export class Controller {
...
+                } else {
+                    let err = await readAll(javaProcess.stderr);
+                    console.log(err);
+                    vscode.window.showErrorMessage("Couldn't determine port app is running on");
+                }
             } else {
-                let err = await readAll(javaProcess.stderr);
-                console.log(err);
-                vscode.window.showErrorMessage("Couldn't determine port app is running on");
+                throw new Error("Couldn't find a JVM to run Java code");
```
Comment (andxu) [inline src/Controller.ts]: report earlier right after findJvm

File: src/Controller.ts (hunk 1 above)
Comment (kdvolder) [reply src/Controller.ts]: I tried something like this before: `import * as opn from "opn";` 

The typescript compiler didn't seem to like that one. When I looked inside the `.d.ts` file for 'opn' it has a comment like this:

```
/* Note that ES6 modules cannot directly export callable functions.
 * This file should be imported using the CommonJS-style:
 *   import x = require('opn');
 */
```

So I just followed that advice and typescript was happy with that, and it seems to work.

If you want me to change it to something else, can you please tell me exactly what you want? Typescript import mechanics aren't 100% understood by me and I'm afraid I'm just not sure what else would work.

File: src/Controller.ts (hunk 3 above)
Comment (kdvolder) [reply src/Controller.ts]: I don't understand what you want me to change here..

File: src/Controller.ts (hunk 4 above)
Comment (kdvolder) [reply src/Controller.ts]: Done.

File: src/Controller.ts (hunk 3 above)
Comment (Eskibear) [reply src/Controller.ts]: I think he want to say:
`vmArgs: vmArgs` can be simply written as `vmArgs` here.

File: src/Controller.ts (hunk 1 above)
Comment (Eskibear) [reply src/Controller.ts]: ES6 is part of the standard now while CommonJS is not, and there're also some slight differences in behavior as far as I know. So usually I use import whenever possible. But here for the package `opn`:

1. By i) uninstall `@types/opn` and ii) remove `"strict": true` from tsconfig.json, I managed to make `import * as opn from "opn"` compiled. 

2. I also tried `import opn = require("opn")`, it works well with `@types/opn`  installed. 

I compared the compiled `Controller.js` files in 1) and 2), and find the two .js file are exactly the same. 

@andxu Can you provide more details and suggestion?

File: src/Controller.ts [hunk 5]
```diff
@@ -7,6 +7,8 @@ import { BootApp, STATE_RUNNING, STATE_INACTIVE } from "./BootApp";
 import { findJvm } from "@pivotal-tools/jvm-launch-utils";
 import * as path from "path";
 import { readAll } from "./stream-util";
+import opn = require('opn')
+import getPort = require('get-port');
```
Comment (Eskibear) [inline src/Controller.ts]: For the package `get-port`, following line works in my machine. 
```typescript
import * as getPort from "get-port";
```

File: src/Controller.ts (hunk 1 above)
Comment (Eskibear) [reply src/Controller.ts]: Missing semicolon in the end of this line, suggest to add it back anyway.

File: src/Controller.ts [hunk 6]
```diff
@@ -38,9 +40,24 @@ export class Controller {
...
+            let jmxport = await(getPort());
+            app.jmxPort = jmxport;
+            let vmArgs = '-Dcom.sun.management.jmxremote ' +
+                `-Dcom.sun.management.jmxremote.port=${jmxport} `+ 
+                '-Dcom.sun.management.jmxremote.authenticate=false ' + 
+                '-Dcom.sun.management.jmxremote.ssl=false ' +
+                '-Djava.rmi.server.hostname=localhost '+ 
+                '-Dspring.application.admin.enabled=true';
+            if (targetConfig.vmArgs) {
+                //TODO: smarter merge? What if user is trying to enable jmx themselves on a specific port they choose, for example?
```
Comment (Eskibear) [inline src/Controller.ts]: This is a good question. I think later we can create an issue to discuss about it.

**Comment by kdvolder** [comment]
After merging branch, I tried it again but it seems I am no longer able to succesfully start my test spring-boot-app, which makes it impossible to also test whether it still opens url.

I beleave this is probably caused somehow by the update in redhat.java (some classpath problem running the app).

**Comment by kdvolder** [comment]
FYI: raised a bug ticket here about my problem launching boot app: https://github.com/Microsoft/vscode-java-debug/issues/402

**Pull request merged**